pkg/store/gcs.go

```diff
@@ -182,151 +200,59 @@ func (s *GCSStore) Close() (err error) {
-	}
-}
-
-func (s *GCSStore) downloadBlocks(ctx context.Context) error {
+func (s *GCSStore) SyncBlocks(ctx context.Context) error {
 	s.metrics.gcsOperations.WithLabelValues(gcsOperationObjectsList).Inc()
+
 	objs := s.bucket.Objects(ctx, &storage.Query{Delimiter: "/"})
 
-	// Fetch a maximum of 20 blocks in parallel.
 	var wg sync.WaitGroup
-	workc := make(chan struct{}, 20)
-
-	for {
-		oi, err := objs.Next()
-		if err == iterator.Done {
-			break
-		} else if err != nil {
-			return err
-		}
-		id, err := ulid.Parse(oi.Prefix[:len(oi.Prefix)-1])
-		if err != nil {
-			continue
-		}
-		if b := s.getBlock(id); b != nil {
-			continue
-		}
-		level.Info(s.logger).Log("msg", "sync block from GCS", "id", id)
+	blockc := make(chan ulid.ULID)
 
+	for i := 0; i < 20; i++ {
 		wg.Add(1)
 		go func() {
-			workc <- struct{}{}
-			s.metrics.blockDownloads.Inc()
+			for id := range blockc {
+				dir := filepath.Join(s.dir, id.String())
 
-			if err := s.downloadBlock(ctx, id); err != nil {
-				level.Error(s.logger).Log("msg", "syncing block failed", "err", err, "block", id.String())
-				s.metrics.blockDownloadsFailed.Inc()
+				b, err := newGCSBlock(ctx, s.logger, id, dir, s.bucket, s.metrics.gcsOperations)
+				if err != nil {
+					level.Warn(s.logger).Log("msg", "loading block failed", "id", id, "err", err)
+					// Wipe the directory so we can cleanly try again later.
+					os.RemoveAll(dir)
+					continue
+				}
+				s.setBlock(id, b)
 			}
 			wg.Done()
-			<-workc
 		}()
 	}
-	wg.Wait()
 
-	return nil
-}
-
-// downloadBlock downloads the index and meta.json file for the given block ID and opens a reader
-// against the block.
-func (s *GCSStore) downloadBlock(ctx context.Context, id ulid.ULID) error {
-	bdir := filepath.Join(s.dir, id.String())
-	tmpdir := bdir + ".tmp"
-
-	if err := os.MkdirAll(tmpdir, 0777); err != nil {
-		return errors.Wrap(err, "create temp dir")
-	}
-
-	for _, fn := range []string{
-		"index",
-		"meta.json",
-	} {
-		obj := s.bucket.Object(path.Join(id.String(), fn))
-
-		f, err := os.Create(filepath.Join(tmpdir, fn))
-		if err != nil {
-			return errors.Wrap(err, "create local index copy")
-		}
-
-		s.metrics.gcsOperations.WithLabelValues(gcsOperationObjectGet).Inc()
-		r, err := obj.NewReader(ctx)
-		if err != nil {
-			return errors.Wrap(err, "create index object reader")
-		}
-		_, copyErr := io.Copy(f, r)
-
-		if err := f.Close(); err != nil {
-			level.Warn(s.logger).Log("msg", "close file", "err", err)
-		}
-		if err := r.Close(); err != nil {
-			level.Warn(s.logger).Log("msg", "close object reader", "err", err)
-		}
-		if copyErr != nil {
-			if err := os.RemoveAll(tmpdir); err != nil {
-				level.Warn(s.logger).Log("msg", "cleanup temp dir after failure", "err", err)
-			}
-			return errors.Wrap(copyErr, "copy index file to disk")
+Loop:
+	for {
+		oi, err := objs.Next()
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			close(blockc)
```

Comment:
No wait for wg group? What about doing just `break` here and making sure we return err